documentation/library_guide/kernel_templates/single_pass_scan.rst

```diff
@@ -0,0 +1,208 @@
+
+The algorithms use global and local device memory (see `SYCL 2020 Specification
+<https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_sycl_device_memory_model>`_)
+for intermediate data storage. For the algorithms to operate correctly, there must be enough memory
+on the device; otherwise, the behavior is undefined. The amount of memory that is required
+depends on input data and configuration parameters, as described below.
+
+Global Memory Requirements
+--------------------------
+
+Global memory is used for copying the input sequence and storing internal data such as status flags.
+The used amount depends on many parameters; below is an upper bound approximation:
+
+2 * V * N \ :sub:`flags` + F * N \ :sub:`flags`
+
+where V is the number of bytes needed to store the input value type and F is the number of bytes needed to store the flags.
+Currently, F is hard-coded to 4 as the flag type is 32-bits.
+
+The value of N\ :sub:`flags` represents the number of work-groups and depends on ``param.data_per_workitem`` and ``param.workgroup_size``.
+It can be approximated by dividing the number of input elements N by the product of ``param.data_per_workitem`` and ``param.workgroup_size``
+and adding 33 for padding.
+
+.. note::
+
+   If the number of input elements can be efficiently processed by a single work-group,
+   the kernel template is executed by a single work-group and does not use any global memory.
+
+
+.. _local-memory:
+
+Local Memory Requirements
+-------------------------
+
+Local memory is used for storing elements of the input that are to be scanned by a single work group.
+The used amount is denoted as N\ :sub:`elems_per_workgroup`, which equals to ``sizeof(key_type) * param.data_per_workitem * param.workgroup_size``.
+
+Some amount of local memory is also used by the calls to SYCL's group reduction and group scan. The amount of memory used particularly
+for these calls is implementation dependent.
+
+-----------------------------------------
+Recommended Settings for Best Performance
+-----------------------------------------
+
+The general advice is to choose kernel parameters based on performance measurements and profiling information.
+The initial configuration may be selected according to these high-level guidelines:
+
+
+- When the number of elements is small enough to fit within single work-group, the algorithm will ignore kernel
+  parameters and instead dispatch to a single workgroup version, where it is generally more efficient.
+
+- Generally, utilizing all available
+  compute cores is key for better performance. To allow sufficient work to satisfy all
+  X\ :sup:`e`-cores [#fnote1]_ on a GPU, use ``param.data_per_workitem * param.workgroup_size ≈ N / xe_core_count``.
+
+- On devices with mutiple tiles, it may prove beneficial to experiment with different tile hierarchies as described
+  in https://www.intel.com/content/www/us/en/developer/articles/technical/flattening-gpu-tile-hierarchy.html.
```

Comment:
Please change this to a live link: `Options for using a GPU Tile Hierarchy <https://www.intel.com/content/www/us/en/developer/articles/technical/flattening-gpu-tile-hierarchy.html>`